pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -486,13 +486,54 @@ func applyConnectionPool(cluster *v2.Cluster, settings *networking.ConnectionPoo
 		if settings.Tcp.MaxConnections > 0 {
 			threshold.MaxConnections = &types.UInt32Value{Value: uint32(settings.Tcp.MaxConnections)}
 		}
+
+		applyTcpKeepalive(env, cluster, settings)
 	}
 
 	cluster.CircuitBreakers = &v2_cluster.CircuitBreakers{
 		Thresholds: []*v2_cluster.CircuitBreakers_Thresholds{threshold},
 	}
 }
 
+func applyTcpKeepalive(env *model.Environment, cluster *v2.Cluster, settings *networking.ConnectionPoolSettings) {
+	var keepaliveProbes uint32
+	var keepaliveTime *types.Duration
+	var keepaliveInterval *types.Duration
+
+	upstreamConnectionOptions := &v2.UpstreamConnectionOptions{
+		TcpKeepalive: &core.TcpKeepalive{},
+	}
```

Comment:
doesn't this mean tcp keepalives are enabled for all clusters irrespective of whether it was enabled globally or locally via destination rules?

Reply:
That's true. This code creates empty tcp_keepalive. Just checked envoy code, has_tcp_keepalive will be true and that sets ENVOY_SOCKET_SO_KEEPALIVE. Will rectify it.